cdm/core/src/test/java/ucar/nc2/util/cache/TestS3NetcdfFileCache.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 1998-2022 University Corporation for Atmospheric Research/Unidata
+ * See LICENSE for license information.
+ */
+package ucar.nc2.util.cache;
```

Comment:
I don't think you can run the S3 tests in this package because in this package you won't be able to open S3 files. I think it will have to be moved to `package ucar.unidata.io.s3;`. This is because our S3 dependencies are not in the cdm core package and the S3RandomAccessFile.Provider only gets dynamically loaded in the S3 package.

Moving these test to another package may require rewriting them a bit because I see some package private functions being used here (like `cache.hits`). If you just want to see what is in the cache that is public (`RandomAccessFile.getGlobalFileCache().showCache()`) so maybe these tests could be simplified. If it is really necessary to check the hits/misses to test this issue we could consider adding a public method to the cache for that.

Reply:
Since the original bug was that the requests were silently missing the cache, I think it is important to have a test that verifies that the cache is performing as expected. This means we need some sort of way to verify the cache state, whether it be by counting the hits and misses, or having a list of the cache keys to test against. I'll leave it up to you devs to decide what to expose as read-only public. I can adjust accordingly.

Good point about being in the wrong location. I'm not used to developing Java libraries, so I am not used to thinking about such considerations.

Reply:
I think actually the existing functions should be usable for your tests (if a bit clunky), so probably no need to add extra functionality there. To see a list of entries in the cache you can do `RandomAccessFile.getGlobalFileCache().showCache()`. To check the stats (as a string like "hits= 0 miss= 0 nfiles= 0 elems= 0") you can something like
```
    final FileCacheIF rafFileCache = RandomAccessFile.getGlobalFileCache();
    final Formatter formatter = new Formatter();
    rafFileCache.showStats(formatter);
    assertThat(formatter.toString()).contains("miss= " + size);
```

Reply:
Great suggestion. I went ahead and redesigned it. Still think it would be a good idea to eventually consider utilizing something like `moto` to  provide a mocked S3 service, but that is obviously outside the scope of this PR.

Reply:
Thanks for the suggestion, I agree that would be nice!